src/python/pants/option/subsystem.py

```diff
@@ -133,6 +136,28 @@ def inner() -> Iterable[Rule]:
 
         return list(inner())
 
+    @final
+    @memoized_classproperty
+    def _plugin_option_cls(cls) -> type:
```

Comment:
Do you think we should comment that the memoization is important for registration to work correctly?

Reply:
Not a bad idea. Since this pattern is used in a few places, I think we can turn this into a normal class property calling into a memoized helper function.

We lose the locality of the memoization, but can centralize the comment/docstring explaining

Reply:
I'm gonna split this into a new PR, since it touches so many places.

Reply:
Oh yeah, I have a really clean solution to this!

Reply:
https://github.com/pantsbuild/pants/pull/17040